### PR TITLE
Parameterize behaviors

### PR DESCRIPTION
**Pull request by vzvzv**

### CODE REVIEW

File: handle_actions.go [hunk 1]
```diff
@@ -33,68 +33,34 @@ func (m *Mock) DoActions(c *Context) error {
...
-		return err
-	}
-
-	if err := m.doActionReplyHTTP(c, a); err != nil {
-		action = "reply_http"
-		return err
-	}
-
-	return nil
+	return actualAction.Perform(*c)
```
Comment (zhouzhuojie) [inline handle_actions.go]: please check if `c != nil` before dereference, otherwise it may cause panic.

File: handle_actions.go [hunk 2]
```diff
@@ -33,68 +33,34 @@ func (m *Mock) DoActions(c *Context) error {
...
 func (m *Mock) doAction(c *Context, a Action) (err error) {
-	var action string
-
+	actualAction := a.GetActualAction()
 	defer func() {
 		if err != nil {
 			logrus.WithFields(logrus.Fields{
 				"err":    err,
-				"action": action,
+				"action": reflect.TypeOf(actualAction).String(),
```
Comment (zhouzhuojie) [inline handle_actions.go]: you can use `%T` to get the type in fmt package. https://golang.org/pkg/fmt/

```
fmt.Sprintf("%T", actualAction)
```

Also, avoid using reflect if possible.

File: handle_actions.go [hunk 3]
```diff
@@ -33,68 +33,34 @@ func (m *Mock) DoActions(c *Context) error {
...
 
-func (m *Mock) doActionSendHTTP(c *Context, a Action) error {
-	if structs.IsZero(a.ActionSendHTTP) {
-		return nil
-	}
-
-	sendHTTP := a.ActionSendHTTP
-
-	bodyStr, err := c.Render(sendHTTP.Body)
+func (actionSendHTTP ActionSendHTTP) Perform(context Context) error {
```
Comment (zhouzhuojie) [inline handle_actions.go]: I'm fine with other naming in the function body, but this receiver `actionSendHTTP` is just too long to read in idiomatic go.

Comment (zhouzhuojie) [inline handle_actions.go]: nit, for `context` go prefers `ctx` as the variable name. `context` is mostly used as the package name. See https://blog.golang.org/context

File: handle_actions.go (hunk 1 above)
Comment (vzvzv) [reply handle_actions.go]: addressed by losing the pointers

File: handle_actions.go (hunk 2 above)
Comment (vzvzv) [reply handle_actions.go]: done 👍

File: handle_actions.go (hunk 3 above)
Comment (vzvzv) [reply handle_actions.go]: you mentioned 'self' was okay?

Comment (vzvzv) [reply handle_actions.go]: should we then also rename our `Context` struct, if that's the confusion we're trying to avoid?

Comment (zhouzhuojie) [reply handle_actions.go]: `Context` is fine, it means a struct; `context` is usually used to refer to the package name from `import "context"`

Comment (gusk) [reply handle_actions.go]: Context should be ok here, implying the package

Comment (zhouzhuojie) [reply handle_actions.go]: definitely ok with me. the community would pick for example `a` in this case.

**Pull request merged**